packages/libs/contracts-sdk/contracts/facets/VincentUserFacet.sol

```diff
@@ -97,137 +94,132 @@ contract VincentUserFacet is VincentBase {
-        // This ensures apps permitted before the allPermittedApps tracking was added are captured
-        us_.agentPkpTokenIdToAgentStorage[pkpTokenId].allPermittedApps.add(appId);
-
-        emit LibVincentUserFacet.AppVersionUnPermitted(pkpTokenId, appId, appVersion);
+        VincentAppStorage.appStorage().appIdToApp[appId].appVersions[getAppVersionIndex(appVersion)].delegatedAgentAddresses
+            .remove(msg.sender);
+
+        // Store the app id, version and pkp signer as last permitted before removing (for potential re-permitting)
+        agentStorage.lastPermittedAppId = appId;
+        agentStorage.lastPermittedAppVersion = appVersion;
+        agentStorage.lastPermittedPkpSigner = agentStorage.pkpSigner;
+        agentStorage.lastPermittedPkpSignerPubKey = agentStorage.pkpSignerPubKey;
+
+        // Remove the App Version from the User's Permitted App
+        agentStorage.permittedAppVersion = 0;
+        agentStorage.permittedAppId = 0;
+        agentStorage.pkpSigner = address(0);
+        agentStorage.pkpSignerPubKey = 0;
+
+        emit LibVincentUserFacet.AppVersionUnPermitted(msg.sender, appId, appVersion, agentStorage.lastPermittedPkpSigner, agentStorage.lastPermittedPkpSignerPubKey);
     }
 
     /**
      * @notice Re-permits a previously unpermitted app with its last permitted version
-     * @dev This function allows a PKP owner to quickly re-authorize an app that was previously unpermitted,
+     * @dev This function allows an agent to quickly re-authorize an app that was previously unpermitted,
      *      using the last version that was permitted. The app must have been previously permitted
-     *      (exists in allPermittedApps) and the last permitted version must still be enabled.
+     *      and the last permitted version must still be enabled.
      *
-     * @param pkpTokenId The token ID of the PKP to re-permit the app for
      * @param appId The ID of the app to re-permit
      */
-    function rePermitApp(uint256 pkpTokenId, uint40 appId) external appNotDeleted(appId) onlyPkpOwner(pkpTokenId) {
+    function rePermitApp(uint40 appId) external appNotDeleted(appId) {
         VincentUserStorage.UserStorage storage us_ = VincentUserStorage.userStorage();
-        VincentUserStorage.AgentStorage storage agentStorage = us_.agentPkpTokenIdToAgentStorage[pkpTokenId];
+        VincentUserStorage.AgentStorage storage agentStorage = us_.agentAddressToAgentStorage[msg.sender];
 
         // Check if app is currently permitted
-        if (agentStorage.permittedApps.contains(appId)) {
+        if (agentStorage.permittedAppId == appId) {
             revert LibVincentUserFacet.AppVersionAlreadyPermitted(
-                pkpTokenId, appId, agentStorage.permittedAppVersion[appId]
+                msg.sender, appId, agentStorage.permittedAppVersion
             );
         }
 
+        if (agentStorage.permittedAppId != 0) {
+            revert LibVincentUserFacet.DifferentAppAlreadyPermitted(msg.sender, appId);
```

Comment:
Misleading error if permitted App ID is `1` and the caller attempt to repermit App ID `1`